crates/sp-domains-fraud-proof/src/host_functions.rs

```diff
@@ -71,6 +80,34 @@ pub trait FraudProofHostFunctions: Send + Sync {
         call_data: &[u8],
         domain_runtime_code: Vec<u8>,
     ) -> Option<Vec<u8>>;
+
+    fn check_extrinsics_in_single_context(
+        &self,
+        domain_runtime_code: Vec<u8>,
+        domain_block_id: (BlockNumber, H256),
```

Comment:
This could be problematic moving between Host and WASM and might have a penatly for huge sizes. Maybe i'm wrong here

Reply:
Do you mean the `domain_runtime_code`? for now, it seems okay to do so in the test and I will test it again with the malicious operator in a more real-world environment. Also, since now we can't fetch the runtime code from the host function while we need to use the runtime code in there it seems we have to pass through the runtime code.

Reply:
> Also, since now we can't fetch the runtime code from the host function

I'm not sure I understand what you meant here. We are currently fetching the domain runtime code within the host function using the consensus hash provided. 

I do not understand where the domain runtime is coming from in the first place. Is it part of the fraud proof or are you fetching this from within the runtime of consensus.

If it is latter, then why cannot we fetch the same within the host function by passing the Consensus hash like we did before.

I do not fully know the performance of moving large chunk of bytes through pass through. While I do like to avoid that, benchmarking would be helpful rather than refactoring it back later.
Any thougths @nazar-pc ?

Reply:
The whole point of stateless FP is to let the FP be self-contained and not rely on the host function to fetch any state because fast-synced farmer may not have these states available and will fail to verify FP or import block (if the block contains FP). While the domain runtime code is part of the consensus state we can't fetch it from the host function.

>I do not understand where the domain runtime is coming from in the first place. Is it part of the fraud proof or are you fetching this from within the runtime of consensus.

Sorry for not making it clear, if the required domain runtime code (i.e.  at `bad_er.consensus_block_number`) is not  upgraded and is still available in the current state then we will just fetch it from the state, otherwise the prover need to generate storage proof for the domain runtime code.

Reply:
> if the required domain runtime code (i.e. at bad_er.consensus_block_number) is not upgraded and is still available in the current state then we will just fetch it from the state

This is exactly what i'm taking about. If Domain runtime code is avaible on Runtime state, then there is not reason to send it through pass through but rather fetch it from the state with in the host function. I'm merely suggesting fetching the same data from runtime within host function instead of sending it through Pass through.

Reply:
How do you guarantee that it will be still available by the time fraud proof is submitted? Can't there be a domain runtime upgrade in between?

Reply:
You are quick to respond. I was suggesting an approach here - https://github.com/subspace/protocol-specs/pull/31#discussion_r1599756311

Why not store the runtime upgrades within the consensus state rather than pruning them? Then fraud proof will only need to carry runtime hash and the latest state of the consensus will be gaurenteed to have that domain runtime.

This will ensure Fraud proof size not unbounded due to domain runtime and there is no reason to pass the Domain runtime through PassThrough

Reply:
Well, looks like there is not necessarily consensus on what needs to be implemented in the first place then :confused:

Reply:
:)
I do not necessarily want to block this current approach but from our past experience and work we did to ensure FP sizes are bounded to guaranteed inclusion in the Consensus block, this approach seems to re-introduce such issue only to be pointed out by either us or audit team regarding the potential unboundedness of the FP down the line.

One approach is to store all runtimes would be trivial change and still keep the FP size bounded but might require more thought into other potential solutions as well

Reply:
Well, this is not the first time we're implementing fraud proofs on my memory. Each time we try to save time during impleentation and end up changing it again. Not sure if it is different this time, btu I'd be unfortunate if we have to significantly change it later and maybe end up being unable to upgade 3h after it lands.

Reply:
Whether we should keep the old domain runtime on the state and not rely on storage proof or not is another thing we can discuss, but for whether the runtime code should be fetched at the host function or pass through from the runtime I'm much like the latter (i.e. the current implementation).

The reason is that the host function is the only source of indeterministic during block execution which we should avoid as much as possible. While the FP verification unfortunately requires host function for something involving the domain runtime, we should still make the host function as deterministic as possible (in this PR most host function just instantiate a domain runtime and make a stateless runtime API call). But fetching state in the host function is something stateful that is related to the node's backend storage, we need to make another effort to ensure it will return a deterministic result for all the node.

For the performance cost of passing data through the host function, I didn't look closer but it seemed just needed to do another encode/decode, and I expected it is something negligible (especially fraud proof is rare) with or without this cost should not make much difference.

Reply:
> The reason is that the host function is the only source of indeterministic during block execution which we should avoid as much as possible.

I dont necessarily agree that usage of host function is indeterministic.
I would avoid this back and forth without actual performance state. Lets move on